nav2_rviz_plugins/src/nav2_panel.cpp

```diff
@@ -48,29 +58,99 @@ Nav2Panel::Nav2Panel(QWidget * parent)
+  QObject::connect(canceled_, SIGNAL(entered()), this, SLOT(onCancel()));
 
   initial_->addTransition(start_stop_button_, SIGNAL(clicked()), starting_);
   starting_->addTransition(start_stop_button_, SIGNAL(clicked()), stopping_);
+  running_->addTransition(start_stop_button_, SIGNAL(clicked()), canceled_);
+  canceled_->addTransition(start_stop_button_, SIGNAL(clicked()), stopping_);
+  completed_->addTransition(start_stop_button_, SIGNAL(clicked()), stopping_);
 
-  machine_.addState(initial_);
-  machine_.addState(starting_);
-  machine_.addState(stopping_);
+  ROSActionQTransition * startupTransition = new ROSActionQTransition(QActionState::INACTIVE);
+  startupTransition->setTargetState(running_);
+  starting_->addTransition(startupTransition);
 
-  machine_.setInitialState(initial_);
-  machine_.start();
+  ROSActionQTransition * canceledTransition = new ROSActionQTransition(QActionState::INACTIVE);
+  canceledTransition->setTargetState(running_);
+  canceled_->addTransition(canceledTransition);
 
-  // Lay out the items in the panel
+  ROSActionQTransition * runningTransition = new ROSActionQTransition(QActionState::ACTIVE);
+  runningTransition->setTargetState(completed_);
+  running_->addTransition(runningTransition);
 
+  ROSActionQTransition * completedTransition = new ROSActionQTransition(QActionState::INACTIVE);
+  completedTransition->setTargetState(running_);
+  completed_->addTransition(completedTransition);
+
+  state_machine_.addState(initial_);
+  state_machine_.addState(starting_);
+  state_machine_.addState(stopping_);
+  state_machine_.addState(running_);
+  state_machine_.addState(canceled_);
+  state_machine_.addState(completed_);
+
+  state_machine_.setInitialState(initial_);
+  state_machine_.start();
+
+  // Lay out the items in the panel
   QVBoxLayout * main_layout = new QVBoxLayout;
   main_layout->addWidget(start_stop_button_);
   main_layout->setContentsMargins(10, 10, 10, 10);
   setLayout(main_layout);
+
+  auto options = rclcpp::NodeOptions().arguments(
+    {"__node:=navigation_dialog_action_client"});
+  client_node_ = std::make_shared<rclcpp::Node>("_", options);
+
+  action_client_ = rclcpp_action::create_client<nav2_msgs::action::NavigateToPose>(client_node_,
+      "NavigateToPose");
+  goal_ = nav2_msgs::action::NavigateToPose::Goal();
+
+  timer_node_ = std::make_shared<rclcpp::Node>("nav2_panel_timer");
+
+  QObject::connect(&GoalUpdater, SIGNAL(updateGoal(double,double,double,QString)),  // NOLINT
+    this, SLOT(onNewGoal(double,double,double,QString)));  // NOLINT
+
+  // Launch a thread to run the node
+  thread_ = std::make_unique<std::thread>(
+    [&](rclcpp::Node::SharedPtr node)
+    {
+      executor_.add_node(node->get_node_base_interface());
+      executor_.spin();
+      executor_.remove_node(node->get_node_base_interface());
+    }, timer_node_);
+
+  timer_ = timer_node_->create_wall_timer(1s, std::bind(&Nav2Panel::timerActionEvent, this));
```

Comment:
You could check more frequently than once per second. This would make it more responsive to detecting the completed navigation at little expense. For example, running the callback at 5Hz should be negligible.